File: find_winners.py

import json
import re

# Complete results from cloud-seeder.js
JORNADAS_RESULTS = {
    1: ["2","1","1","2","X","2","1","1","1","X","2","1","X","X","1-1"],
    2: ["X","X","2","1","X","1","2","1","2","2","2","X","2","2","1-2"],
    3: ["X","1","2","1","X","2","1","1","X","X","1","1","2","2","1-1"],
    5: ["X","1","2","1","2","1","1","1","1","X","1","1","X","X","1"],
    7: ["2","1","1","1","2","X","X","1","1","1","X","1","1","2","2-0"],
    9: ["X","1","1","2","1","1","1","1","X","2","2","1","X","2","M-2"],
    11: ["2","1","1","1","1","1","2","2","X","2","1","2","1","1","1-1"],
    14: ["2","1","X","1","X","X","2","2","2","1","2","2","1","X","0-0"],
    16: ["X","1","2","2","X","2","1","2","X","2","1","1","1","X","2-1"],
    17: ["1","1","1","2","1","1","1","X","2","1","X","X","1","2","M-2"],
    19: ["1","1","1","2","1","X","1","2","X","X","1","X","1","X","0-0"],
    22: ["2","1","2","1","X","X","2","X","2","X","1","1","2","1","2-1"],
    24: ["X","1","2","1","2","2","2","X","1","2","X","1","2","1","1-1"],
    26: ["1","1","2","1","X","","","","2","2","2","X","","","1-0"]
}

MEMBERS = [
    "Alvaro", "Carlos", "David Buzón", "Edu", "Emilio",
    "Fernando Lozano", "Fernando Ramírez", "Heradio", "JA Valdivieso", "Javier Mora",
    "Juan Antonio", "Juanjo", "Luismi", "Marcelo", "Martín",
    "Rafa", "Ramón", "Raúl Romera", "Samuel"
]

def calculate_aciertos(official, forecast):
    if not official or not forecast: return 0
    hits = 0
    for i in range(min(14, len(official), len(forecast))):
        if official[i] and forecast[i] and str(official[i]) == str(forecast[i]):
            hits += 1
    return hits

with open('d:/PROYECTO_MAULAS/js/data_loader.js', 'r', encoding='utf-8') as f:
    lines = f.readlines()
    json_lines = []
    capture = False
    for line in lines:
        if 'window.HISTORICAL_DATA = [' in line:
            capture = True
            json_lines.append('[')
            continue
        if capture:
            if line.strip() == '];':
                json_lines.append(']')
                capture = False
                break
            json_lines.append(line)
    
    json_str = "".join(json_lines)
    data = json.loads(json_str)

winners = {}
for h in data:
    j_num = h['jornada_num']
    official = JORNADAS_RESULTS.get(j_num)
    if not official: continue

    ranking = []
    for m_id, preds in h['predictions'].items():
        hits = calculate_aciertos(official, preds)
        ranking.append({
            'mId': int(m_id),
            'name': MEMBERS[int(m_id)-1],
            'hits': hits
        })
    
    ranking.sort(key=lambda x: (-x['hits'], x['mId']))
    winners[j_num] = ranking

# Generate the data report
target_jornadas = [1, 2, 8, 10, 13, 15, 16, 18, 21, 23, 25, 27, 31]
for j_target in sorted([1, 2, 3, 9, 11, 14, 16, 17, 19, 22, 24, 26, 28, 32]):
    j_prev = j_target - 1
    if j_target == 1:
        print(f"J1: Luismi (ID 13) - Maula J0")
        continue
    
    if j_prev in winners:
        w = winners[j_prev]
        print(f"J{j_target}: Winner J{j_prev}: {w[0]['name']} (ID {w[0]['mId']}) - {w[0]['hits']} hits")
        if j_target in [3, 9]:
            print(f"J{j_target}: 2nd J{j_prev}: {w[1]['name']} (ID {w[1]['mId']}) - {w[1]['hits']} hits")
    else:
        print(f"J{j_target}: Prev Jornada J{j_prev} results missing in seeder.")

File: find_winners.js

const fs = require('fs');
const path = require('path');

// Mock a minimal environment to run the logic
const window = { AppUtils: { parseDate: (d) => new Date(d) } };

// Load data_loader.js content (I'll need to strips 'window.HISTORICAL_DATA =' part)
const dataLoaderContent = fs.readFileSync('d:/PROYECTO_MAULAS/js/data_loader.js', 'utf8');
const HISTORICAL_DATA = eval(dataLoaderContent.replace('window.HISTORICAL_DATA =', ''));

// Results from cloud-seeder.js (I'll extract them manually to keep script simple)
const JORNADAS_RESULTS = {
    1: ["2", "1", "1", "2", "X", "2", "1", "1", "1", "X", "2", "1", "X", "X", "1-1"],
    2: ["X", "X", "2", "1", "X", "1", "2", "1", "2", "2", "2", "X", "2", "2", "1-2"],
    3: ["X", "1", "2", "1", "X", "2", "1", "1", "X", "X", "1", "1", "2", "2", "1-1"],
    4: null, // missing?
    5: ["X", "1", "2", "1", "2", "1", "1", "1", "1", "X", "1", "1", "X", "X", "1"],
    6: null,
    7: ["2", "1", "1", "1", "2", "X", "X", "1", "1", "1", "X", "1", "1", "2", "2-0"],
    8: null,
    9: ["X", "1", "1", "2", "1", "1", "1", "1", "X", "2", "2", "1", "X", "2", "M-2"],
    10: null,
    11: ["2", "1", "1", "1", "1", "1", "2", "2", "X", "2", "1", "2", "1", "1", "1-1"],
    12: null,
    13: null,
    14: ["2", "1", "X", "1", "X", "X", "2", "2", "2", "1", "2", "2", "1", "X", "0-0"],
    15: null,
    16: ["X", "1", "2", "2", "X", "2", "1", "2", "X", "2", "1", "1", "1", "X", "2-1"],
    17: ["1", "1", "1", "2", "1", "1", "1", "X", "2", "1", "X", "X", "1", "2", "M-2"],
    18: null,
    19: ["1", "1", "1", "2", "1", "X", "1", "2", "X", "X", "1", "X", "1", "X", "0-0"],
    20: null,
    21: null,
    22: ["2", "1", "2", "1", "X", "X", "2", "X", "2", "X", "1", "1", "2", "1", "2-1"],
    23: null,
    24: ["X", "1", "2", "1", "2", "2", "2", "X", "1", "2", "X", "1", "2", "1", "1-1"],
    25: null,
    26: ["1", "1", "2", "1", "X", "", "", "", "2", "2", "2", "X", "", "", "1-0"]
};

// Members mapping
const members = [
    "Alvaro", "Carlos", "David Buzón", "Edu", "Emilio",
    "Fernando Lozano", "Fernando Ramírez", "Heradio", "JA Valdivieso", "Javier Mora",
    "Juan Antonio", "Juanjo", "Luismi", "Marcelo", "Martín",
    "Rafa", "Ramón", "Raúl Romera", "Samuel"
];

function calculateAciertos(official, forecast) {
    if (!official || !forecast) return 0;
    let aciertos = 0;
    for (let i = 0; i < 14; i++) {
        if (official[i] && forecast[i] && official[i].toString() === forecast[i].toString()) {
            aciertos++;
        }
    }
    return aciertos;
}

const winners = {};
HISTORICAL_DATA.forEach(h => {
    const jNum = h.jornada_num;
    const official = JORNADAS_RESULTS[jNum];
    if (!official) return;

    const ranking = [];
    Object.entries(h.predictions).forEach(([mId, preds]) => {
        const hits = calculateAciertos(official, preds);
        ranking.push({ mId: parseInt(mId), name: members[mId - 1], hits });
    });

    ranking.sort((a, b) => b.hits - a.hits || a.mId - b.mId);
    winners[jNum] = ranking;
});

console.log(JSON.stringify(winners, null, 2));
